### ISSUE TITLE
SCRAM-SHA-1(-PLUS) + SCRAM-SHA-256(-PLUS) supports

### ISSUE THREAD

**Issue by Neustradamus** [jasonmunro/cypht#354]
"When using the SASL SCRAM mechanism, the SCRAM-SHA-256-PLUS variant SHOULD be preferred over the SCRAM-SHA-256 variant, and SHA-256 variants [RFC7677] SHOULD be preferred over SHA-1 variants [RFC5802]".

Can you add support for?
- SCRAM-SHA-1(-PLUS):
-- https://tools.ietf.org/html/rfc5802
-- https://tools.ietf.org/html/rfc6120

- SCRAM-SHA-256(-PLUS):
-- https://tools.ietf.org/html/rfc7677 since 2015-11-02
-- https://tools.ietf.org/html/rfc8600 since 2019-06-21: https://mailarchive.ietf.org/arch/msg/ietf-announce/suJMmeMhuAOmGn_PJYgX5Vm8lNA

I add SCRAM-SHA-512(-PLUS): https://xmpp.org/extensions/inbox/hash-recommendations.html

Linked to:
- https://github.com/scram-xmpp/info/issues/1

**Comment by jasonmunro** [comment]
I will look into this, thanks for the report.

**Comment by Neustradamus** [comment]
Dear @cypht-org team,

Have you progressed since 2019?

Thanks in advance.

**Comment by marclaporte** [comment]
Hi @Neustradamus 

Thank you for the follow up. AFAIK, no progress has been made on this specific issue.

But we've just released a major version with tona of new features and fixes. And we've moved the project to an organization to facilitate work by various developers. Ref.: https://unencumberedbyfacts.com/2023/06/14/cypht-rebooted/

@josaphatim What do you think?

**Comment by Neustradamus** [comment]
@marclaporte: Excellent news!

Hope it will be possible to have more security in your other projects too :)

**Comment by marclaporte** [comment]
Hehe, yeah :-)

**Comment by marclaporte** [comment]
@Danelif please advise.

**Comment by Danelif** [comment]
I will look into this and give report

**Comment by Danelif** [comment]
@marclaporte I have read all the documentation provided above and in my opinion crypt SCRAM-SHA-256-PLUS is suitable. Unfortunately, there is no many documentations apart from Internet Engineering Task Force (IETF). 
but I found this video interesting  https://www.youtube.com/watch?v=20p4zP_pvQU&pp=ugMICgJmchABGAHKBSxwcm90ZWdlciB2b3MgbW90cyBkZSBwYXNzZSBzY3JhbSBwb3N0Z3Jlc3FsIA%3D%3D

**Comment by Neustradamus** [comment]
@Danelif: Yes, PostgreSQL supports SCRAM-SHA-256 and SCRAM-SHA-256-PLUS and a lot of others projects/softs/libs/...
There is a list here: https://github.com/scram-sasl/info/issues/1.

@marclaporte: I have tried to contact you in private by e-mail, by Twitter (public message because DM is not opened for you), can you reply me? Thanks in advance.

**Comment by marclaporte** [comment]
> can you reply me

Done

**Comment by marclaporte** [comment]
@Danelif Do you feel comfortable to prepare a pull request?

**Comment by Danelif** [comment]
@marclaporte yes I do

**Comment by marclaporte** [comment]
ok, please proceed

**Comment by Neustradamus** [comment]
@marclaporte, @Danelif: Can you look SCRAM with Auth_SASL/Auth_SASL2?
- https://github.com/pear/Auth_SASL
- https://github.com/pear/Auth_SASL2

It will be nice to have before a new release...

**Comment by Danelif** [comment]
@Neustradamus okay let me take a look

**Comment by Danelif** [comment]
[Adding SCRAM-256-PLUS PR](https://github.com/cypht-org/cypht/pull/1072) @Neustradamus

**Comment by Neustradamus** [comment]
@Danelif: Nice, good job!

Can you add other SCRAM too?

**Comment by Danelif** [comment]
@Neustradamus alright

**Comment by Neustradamus** [comment]
@Danelif: Note that there are two connection possibilities by SCRAM, example for 256:
- SCRAM-SHA-256
- SCRAM-SHA-256-PLUS

It will be perfect to have support of: SHA-1/SHA-256/SHA-512/SHA3-512.

Note: If you can add the -PLUS variant support into Auth_SASL/Auth_SASL2, it will be a big enhancement to have a full compatibility and will profit to several projects in the World.

**Comment by Danelif** [comment]
i just finished to include SCRAM-SHA-256

**Comment by marclaporte** [comment]
Yes

**Comment by marclaporte** [comment]
Wait. Is there a use case that the more advanced mechanisms would not be used?

If system always uses more secure mechanisms, it should never fall back to CRAM-MD5

@kroky @Neustradamus

**Comment by marclaporte** [comment]
Say the mail server only supports plain and CRAM-MD5, it's surely better to use CRAM-MD5 instead of plain?

**Comment by Neustradamus** [comment]
At the same time of CRAM-MD5 removal, it is possible to remove LOGIN too.

The latest possibility for "old servers" is PLAIN with TLS.

**Comment by kroky** [comment]
I'd deprecate and use as a last resort for now. If big clients/servers haven't removed it yet, there might be a strong reason to. I understand the protocol is outdated and moved to a historic state but completely removing the authentication protocols might actually drop support for certain servers/users/use-cases.

**Comment by Neustradamus** [comment]
Linked to:
- https://github.com/scram-sasl/info/issues/1
- https://github.com/cypht-org/cypht/pull/1072
- https://github.com/cypht-org/cypht/pull/1082
- https://github.com/cypht-org/cypht/pull/1083
